Metadata API: preserve Role.keyids order

MVrachev opened the pull request:
Fixes #1478 

We made Role.keyids a set because the keyids are supposed
to be unique and this still makes sense.

However, the data should also preserve order
(when deserialized and serialized) and currently, it does not.
This is fairly serious since writing signed data potentially modifies
the data (making the signature invalid).

The simplest solution (as proposed by Jussi) is to make keyids
a property and there to enforce keyids uniqueness and preserve the
order by using a list instead of a set.


**Description of the changes being introduced by the pull request**:

**Please verify and check that the pull request fulfills the following
requirements**:

- [x] The code follows the [Code Style Guidelines](https://github.com/secure-systems-lab/code-style-guidelines#code-style-guidelines)
- [x] Tests have been added for the bug fix or new feature
- [x] Docs have been added for the bug fix or new feature

jku commented on tuf/api/metadata.py:
```diff
@@ -626,15 +637,18 @@ def to_dict(self) -> Dict[str, Any]:
-        self.keys[key.keyid] = key
+        if key.keyid not in self.roles[role].keyids:
+            self.roles[role].keyids.append(key.keyid)
+            self.keys[key.keyid] = key
```
this changes what happens when keyid is in keyids already (keys does not get updated)

jku commented on tuf/api/metadata.py:
```diff
@@ -626,15 +637,18 @@ def to_dict(self) -> Dict[str, Any]:
             KeyError: If 'role' does not include the key
         """
+        if keyid not in self.roles[role].keyids:
+            raise KeyError(f"Role.keyid doesn't include {keyid}!")
```
I guess keyids.remove() doesn't raise KeyError now? 

I think we could just change the API doc instead so whatever is raised is documented

jku commented on tuf/api/metadata.py:
```diff
@@ -524,21 +524,32 @@ class Role:
+                raise ValueError(
+                    f"keyids should be a list of unique strings,"
+                    f" instead got {input_keyids}"
+                )
```
has to be a way to squeeze this into something less than four lines
```suggestion
                raise ValueError(f"Non-unique keyids: {input_keyids}"
```

jku commented:
@MVrachev can you please comment on Teodoras suggestion (comment was made in the issue)? 

I think the idea was that we could keep using the set if we make sure that on serialization we always sort the the output. This has some corner cases as well (like if your using the API to process keyids somehow, the order isn't guaranteed) but if that's the only issue then maybe that's a better way?

MVrachev commented:
Yes, I started answering, but it seems I haven't sent my response. 
It sounds reasonable and it won't add complexity with the additional property and checks that have to be made in my current suggestions.
I will update the pr with those suggestions.

Commit:
Metadata API: preserve Role.keyids order

We made Role.keyids a set because the keyids are supposed
to be unique and this still makes sense.

However, the data should also preserve order
(when deserialized and serialized) and currently, it does not.
This is fairly serious since writing signed data potentially modifies
the data (making the signature invalid).

The simplest solution (as proposed by Teodora) is to sort the
set during serialization and that would ensure the order of the items.

Signed-off-by: Martin Vrachev <mvrachev@vmware.com>

MVrachev commented:
Yes, it looks a lot better.
Great suggestion @sechkova!

jku reviewed (approved):
I'm trying to think if there's any case where in the API we would need the ordered keyids... but I don't think we ever do. So this looks right to me, thanks!
